api/nodes/serializers.py

```
@@ -130,6 +131,128 @@ def update(self, instance, validated_data):
+        if user in node.contributors:
+            raise exceptions.ValidationError('{} is already a validated contributor'.format(user.username))
+
+        bibliographic = validated_data['bibliographic']
+        permissions = self.get_permissions_list(validated_data['permission']) \
+            if 'permission' in validated_data else ['read', 'write']
+        node.add_contributor(contributor=user, auth=auth, visible=bibliographic, permissions=permissions, save=True)
+        node.reload()
+        user.permission = node.get_permissions(user)[-1]
+        user.bibliographic = node.get_visible(user)
+        user.node_id = node._id
+        return user
+
+    @staticmethod
+    def get_permissions_list(permission):
+        if permission == 'admin':
+            return ['read', 'write', 'admin']
+        elif permission == 'write':
+            return ['read', 'write']
+        elif permission == 'read':
+            return ['read']
+        else:
+            return ['read', 'write']
+
+
+class NodeContributorDetailSerializer(NodeContributorsSerializer):
+    """ Overrides node contributor serializer to make id read only and add addtional methods
+    """
+    id = ser.CharField(read_only=True, source='_id')
+
+    def update(self, instance, validated_data):
+        user = instance
+        auth = Auth(self.context['request'].user)
+        node = self.context['view'].get_node()
+        if 'permission' in validated_data and validated_data['permission'] != '':
+            self.set_user_permissions(node, validated_data['permission'], user, auth)
+        try:
+            node.set_visible(user, validated_data['bibliographic'], save=True)
+        except ValueError as e:
+            raise exceptions.ValidationError(e)
+        user.permission = node.get_permissions(user)[-1]
+        user.bibliographic = node.get_visible(user)
+        user.node_id = node._id
+        return user
+
+    # checks to make sure unique admin is removing own admin privilege
```

Hmm. If you get a return of false, could you count the number of contributors and return one error if there's only one and the other if there's more than one? That could save you a bunch of code and complication. But I agree, it should probably raise an exception here.